[RFC] ns-3: init at 28.0

teto opened the pull request:
ns-3 is a discrete event C++ network simulator developed at www.nsnam.org popular in the networking research community.

Python bindings might not be functional yet, I could not test them as the required libraries are still in PR https://github.com/NixOS/nixpkgs/pull/30473 but I have yet to meet someone using the bindings.
The manual building doesn't complete yet but it's mostly useful for core developers upon release.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Mic92 commented on pkgs/development/libraries/science/networking/ns3/default.nix:
```diff
@@ -0,0 +1,87 @@
+# All modules can be enabled by choosing 'all_modules'.
+# included here the DCE mandatory ones
+, modules ? [ "core" "network" "internet" "point-to-point"]
+, pkgs
```
Please do not add `pkgs` and pass `gcc6`, `python` and `castxml` directly here.

teto replied:
you already warned me on that for other PRs. Sorry. I used to start writing the packages  from the overlay, hence the pkgs remaining from "super.pkgs" which I forgot to remove. It's more convenient then writing `callPackage ../ns3 { inherit (super.pkgs) gcc6 python castxml }`. In fact I am curious how people create their package while waiting for a merge. If I add it to nixpkgs, then I lose the package once switching  to another branch so it's hard to develop several packages at the same time.

Mic92 replied:
I personally use a nixpkgs checkout instead of channels: http://anderspapitto.com/posts/2015-11-01-nixos-with-local-nixpkgs-checkout.html

Mic92 commented on pkgs/development/libraries/science/networking/ns3/default.nix:
```diff
@@ -0,0 +1,87 @@
+, withExamples ? false
+
+# --enable-tests
+, withTests ? true
```
I would not make testing optional here. This is done by hydra, so most people will not care.

teto replied:
Agreed

Mic92 commented:
If the manual does not build yet, please do not advertise this as an option.

teto commented on pkgs/development/libraries/science/networking/ns3/default.nix:
```diff
@@ -0,0 +1,87 @@
+   texlive.combined.scheme-medium ];
+
+  propagatedBuildInputs = with python.pythonPackages; with pkgs; [ gcc6 python ]
+    ++ stdenv.lib.optional withGsl [ gsl-bin libgsl2 libgsl ]
```
ns3 supports many optional libraries and configurations, i prefer to keep the build small rather than list all of them. I made an exception for this one because it's supposed to improve simulation speed but I wouldn't mind removing it altogether.

Mic92 replied:
If it increase speed, then I would enable it.

teto replied:
I asked on the channel and in fact it seems to just increase the wifi precision according to https://www.nsnam.org/wiki/Installation#Installation and not necessarily increase speed. I fixed the package names (locally), just gsl was enough

teto commented on pkgs/development/libraries/science/networking/ns3/default.nix:
```diff
@@ -0,0 +1,87 @@
+, dia, tetex ? null, ghostscript ? null, texlive ? null
+
+, withDoc ? false
+, withManual ? false
```
not sure how that will work but I'll look into it. Like if you don't call $outputDoc, it won't be built (or ns-3.doc) ?

Mic92 replied:
you can opt-out from installing documentation.

teto committed:
ns-3: init at 28.0

ns-3 is a discrete event C++ network simulator developed at www.nsnam.org .

Python bindings & manual building are disabled by default but you can
override it.

teto commented:
While building the manual I meet
```
(/nix/store/4agl46x083dpanpbsdydbzpqy1frgd66-texlive-combined-medium-2016/share
/texmf/tex/generic/babel/babel.def)))
(/nix/store/4agl46x083dpanpbsdydbzpqy1frgd66-texlive-combined-medium-2016/share
/texmf/tex/latex/psnfss/times.sty)

! LaTeX Error: File `fncychap.sty' not found.
```
which apparently belongs to
```
nix-locate fncychap.sty                                                                                           ~/nixpkgs
python36Packages.sphinx_1_2.out                  18,775 r /nix/store/1sydasckcwpwnm635mcjkm9g1r36958n-python3.6-sphinx-1.2.3/lib/python3.6/site-packages/sphinx/texinputs/fncychap.sty
(arelle.doc)                                     18,775 r /nix/store/p829jnd9jvhismkyf8b0ja0qylpcyyyp-python3.4-sphinx-1.2.3/lib/python3.4/site-packages/sphinx/texinputs/fncychap.sty
python27Packages.sphinx_1_2.out                  18,775 r /nix/store/n7cdj6jmjf5vhvq32r4v05m097hgfc1n-python2.7-sphinx-1.2.3/lib/python2.7/site-packages/sphinx/texinputs/fncychap.sty
python27Packages.gateone.out                     18,775 r /nix/store/f6k1gcn281xgw1nipqq5fc26sqrny524-python2.7-gateone-1.2-0d57c3/gateone/docs/build/latex/fncychap.sty
python27Packages.gateone.out                     18,775 r /nix/store/f6k1gcn281xgw1nipqq5fc26sqrny524-python2.7-gateone-1.2-0d57c3/lib/python2.7/site-packages/gateone/docs/build/latex/fncychap.sty
```
I wonder where the problem lies: should sphinx or ns3 export some value for latex to find teh files ? maybe the problem has already been met

teto commented:
I removed the gsl part as it' not about speed but about precision and this is not used in core so I had not tested it.
I added the --profile switch so that people can build in release mode (ns3 default is debug)
People should instead override the package to use their own library. I suspect there is not much value in having this in binary cache as ns3 users are likely to override the defaults buildInputs and so on. I think I could add the attribute to enforce local build.

As for the options that don't work, I moved them to a ` let ... in `so that the current progress is not lost.
```
+let
+  # there are various problems with manual generation see PR #31346
+  # once those are solved, promote it to an option
+  withManual = false;
+
+  # required packages are not in mainline yet
+  generateBindings = false;
+in
```

FRidh commented:
> As for the options that don't work, I moved them to a let ... inso that the current progress is not lost.

If it doesn't work, then it should not go in. You can keep the relevant changes in your own branch.

FRidh commented on pkgs/top-level/all-packages.nix:
```diff
@@ -19014,6 +19014,10 @@ with pkgs;
   megam = callPackage ../applications/science/misc/megam { };
 
+  ns-3 = callPackage ../development/libraries/science/networking/ns3 {
+    python=python3;
```
Is Python 3 required? If so, make the parameter `python3` instead of `python`. is it compatible with 2 and 3? Then just use `python`.

teto commented:
I updated the package for ns3.28. Bindings and manual now work (though I disabled them by default as it can be pretty expansive to compile stuff).

Mic92 commented:
Thanks!
